### PR TITLE
Documentation: add a more detailed guide for react native development

### PR DESCRIPTION
**Pull request by gwwar**
This PR adds some clearer first time setup instructions for react native development. I'd also be open to moving this to a separate how-to-guide, if we'd like to keep the current doc brief.

I also have images and video assets I can link to (really necessary for some of the android studio screens) but I'd be interested in knowing where the best place to add this would be. I do see an [assets](https://github.com/WordPress/gutenberg/tree/trunk/docs/assets) folder, but we probably don't want to add image/videos to the repository to avoid bloating repository size.

### Testing Instructions

- Verify that instructions are clear
- If you don't have native development going locally, give the instructions a try!

### CODE REVIEW

**Comment by hypest** [comment]
Thanks for the ping @gziolo and yeah, thanks for opening this ticket @gwwar !

Let me reel in @SiobhyB that has some recent interest in the topic. 👍

**Comment by gwwar** [comment]
So chatting with @mkaz it sounds like it's not ideal to add image/video assets to the Gutenberg repository (since it'll bloat the repository). One way we can work around this is to add these assets to the media library of one of the WordPress.org blogs.

Let me know if folks prefer keeping this as one setup doc, or if it'd be more ideal to make a separate how to article. And of course if there are simpler steps or if I mis-noted anything! For Android setup in particular it took me quite a while to figure out the right path arguments.

**Comment by SiobhyB** [comment]
Wow, this is all so detailed and awesome, thank you for the time putting it together @gwwar 🙇‍♀️ 

I like your thoughts on splitting this off into a separate doc. It feels a bit hard to parse at the length it is now and there's a chance not everyone getting set up is going to need these steps (for example, a new contributor already working with React Native will probably have gone through them). We'd also want to link to these detailed steps from the Gutenberg Mobile repository's README. 

So, all in all, a separate doc with _only_ these steps seems like the best move to me. If you agree with that reasoning, let me know if I can help with any part of separating this out and I'll be happy to review again when it's done :)

**Comment by gwwar** [comment]
Reserving this comment for assets. Folks are welcome to contribute better quality ones. If folks agree, we'll want to add these to the media library of one of the wp.org blogs, then link to them in the how to guide.

cleanshot-2021-10-07-at-15 32 34402x

cleanshot-2021-10-07-at-15 20 02402x

https://user-images.githubusercontent.com/1270189/137405310-765ef960-acca-49ed-a2eb-b87b6bb970e9.mp4

package-manager

show package details

screen-shot-2021-04-29-at-3 25 57-pm

screen-shot-2021-04-29-at-4 53 19-pm

screen-shot-2021-04-29-at-4 56 24-pm

avd

screen-shot-2021-04-30-at-8 16 03-am

screen-shot-2021-04-30-at-8 16 53-am

screen-shot-2021-04-30-at-8 17 09-am

screen-shot-2021-04-30-at-8 20 58-am

https://user-images.githubusercontent.com/1270189/137403245-8e13acc8-27fa-4876-a52e-d497ab65a29d.mp4

screen-shot-2021-04-30-at-8 41 54-am

https://user-images.githubusercontent.com/1270189/137403353-2a8ded47-5c7c-4f99-b2cc-fa6def4b4990.mp4

screen-shot-2021-04-30-at-8 50 11-am

android-e2e

**Comment by gwwar** [comment]
@SiobhyB sounds good, I've moved instructions into a how-to guide in https://github.com/WordPress/gutenberg/pull/35529/commits/aea287c9808db4f4af7094dc4d3312009e9319b9

Let me know if any of the instructions are incorrect or if we should clarify wording!

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md [hunk 1]
```diff
@@ -0,0 +1,255 @@
...
+
+[Add screenshot of project settings]
+
+### Update Paths
+
+Export the following env variables and update $PATH. We can normally add this to our ` ~/.zshrc` file if we're using zsh
+in our terminal, or `.bash_profile` if the terminal is still using bash.
+
+```sh
+export JAVA_HOME=/Applications/Android\ Studio.app/Contents/jre/jdk/Contents/Home
```
Comment (gwwar) [inline docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: Reminder to note that this changed to a slightly different path in Arctic Fox
```
export JAVA_HOME="/Applications/Android Studio.app/Contents/jre/Contents/Home"
```

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md [hunk 2]
```diff
@@ -0,0 +1,255 @@
...
+
+### Putting it all together
+
+Start metro:
+
+```
+npm run native start:reset
+```
+
+In another terminal run the following. The emulator doesn’t need to be launched previously.
```
Comment (SiobhyB) [inline docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]:
```suggestion
In another terminal run the following to launch the demo app in the Android emulator (if the emulator isn't already running, it'll also be launched as part of this command):
```
How about tweaking the wording here? I wasn't clear on the mention of the emulator at first.

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md [hunk 3]
```diff
@@ -0,0 +1,255 @@
...
+[react native doctor](https://reactnative.dev/blog/2019/11/18/react-native-doctor). From your checkout, or
+relative to `/packages/react-native-editor folder`, run:
+
+```sh
+npx @react-native-community/cli doctor
+```
+
+[Add screenshot of cli doctor]
+
+See if doctor can fix both "common" and "iOS" issues. At this stage "Android" will still have ❌s for items.
```
Comment (SiobhyB) [inline docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]:
```suggestion
See if `doctor` can fix both "common" and "iOS" issues. (Don't worry if "Android" still has ❌s at this stage, we'll get to those later!)
```
As we can't be certain that the reader won't have their Android set up ready at this point, how about the tweaked wording to account for that?

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md [hunk 4]
```diff
@@ -0,0 +1,255 @@
...
+[Add screenshot of project settings]
+
+### Update Paths
+
+Export the following env variables and update $PATH. We can normally add this to our ` ~/.zshrc` file if we're using zsh
+in our terminal, or `.bash_profile` if the terminal is still using bash.
+
+```sh
+export JAVA_HOME=/Applications/Android\ Studio.app/Contents/jre/jdk/Contents/Home
+export ANDROID_HOME=/Users/{your-username}/Library/Android/sdk
```
Comment (SiobhyB) [inline docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: @gwwar, I'm curious whether the following paths also work for you? These are what I have in my `~/.zshrc` file, which I got from [the React Native docs:](https://reactnative.dev/docs/environment-setup):

```
export ANDROID_HOME=$HOME/Android/Sdk
export PATH=$PATH:$ANDROID_HOME/emulator
export PATH=$PATH:$ANDROID_HOME/tools
export PATH=$PATH:$ANDROID_HOME/tools/bin
export PATH=$PATH:$ANDROID_HOME/platform-tools
```
As those are directly from the official docs, I'd lean towards using them, but let me know if they don't work as expected for you.

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md [hunk 5]
```diff
@@ -0,0 +1,255 @@
+# A first-time setup guide to native mobile development (OSX)
+
+Are you interested in contributing to native mobile development? This
+guide is a detailed walk through designed to get you up and running!
+
+Note that the following instructions here are primarily focused on the
+OSX environment.
```
Comment (SiobhyB) [inline docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]:
```suggestion
OSX environment. For other environments, [the React Native quickstart documentation](https://reactnative.dev/docs/environment-setup) has helpful pointers and steps for getting set up.
```
How about pointing to the RN doc here in case someone needs guidance for other environments?

**Review by SiobhyB — CHANGES_REQUESTED** [review]
Thanks again for all the work on this @gwwar, I've added a few small comments/suggestions throughout, but it generally all looks excellent 🙇‍♀️

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md (hunk 4 above)
Comment (gwwar) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: They don't, but let me see if there's a way to consistently install in that directory.

Also is there a smarter way to install needed packages? I had to lookup needed items in the build.gradle file.

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md (hunk 5 above)
Comment (gwwar) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: Updated in https://github.com/WordPress/gutenberg/pull/35529/commits/ada0609a2cc4568a1d4c51b7fb75c48c4ebe583d

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md (hunk 3 above)
Comment (gwwar) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: Updated in https://github.com/WordPress/gutenberg/pull/35529/commits/ada0609a2cc4568a1d4c51b7fb75c48c4ebe583d

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md (hunk 2 above)
Comment (gwwar) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: Changes read well to me, updated in https://github.com/WordPress/gutenberg/pull/35529/commits/ada0609a2cc4568a1d4c51b7fb75c48c4ebe583d

File: docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md (hunk 4 above)
Comment (gwwar) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: So I do like the updated paths relative to `$HOME` 👍  I left a note inline on where the SDK home location can be modified. By default a new install on OSX will recommend `$HOME/Library/Android/sdk`

I also had to set JAVA_HOME. I pointed this at what comes with Android studio which is convenient. If there's a better alternative, I'm open to changes.

One other note is that I have a working environment on my Intel Mac but not my new M1. Currently getting stuck here with this trace:
CleanShot 2021-10-25 at 15 41 53@2x

Comment (SiobhyB) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: > Also is there a smarter way to install needed packages? I had to lookup needed items in the build.gradle file.

I'm afraid the only way I'm aware of is manually looking up the needed packages (I did that through [the docs](https://reactnative.dev/docs/environment-setup) in my case, but your approach is also valid). I had a quick search to see if there were perhaps any Android Studio plugins to automate the process, but wasn't able to find anything. 

> One other note is that I have a working environment on my Intel Mac but not my new M1. Currently getting stuck here with this trace

@twstokes also ran into some problems running the demo app on an M1 and found he needed to install Platform S packages. These were the specific steps here shared:  

* In Android Studio, go to Preferences -> Appearance & Behavior -> System Settings -> Android SDK.
* Check “Show Package Details” at the bottom right.
* If not already selected, under “Android S Preview” check:
  * Android SDK Platform S
  * Google APIs ARM 64
  * Google Play ARM 64
* Click OK.

And a screenshot of his settings:

screen-shot-2021-07-03-at-12 58 02

Another issue that Tanner came across is a bug that prevents the Android emulator from loading. This was in May, so hopefully, there will have been a fix by now. If not, Tanner followed the workaround outlined in [this Medium article](https://medium.com/mobile-app-development-publication/m1-android-emulator-not-detected-workaround-c7718f08ba99). 

I don't have an M1 but, if installing the Platform S packages doesn't get things working for you, we could see if Tanner or others on the project have further ideas here. Let me know how it goes! (and thank you again for all the work documenting this!)

Comment (SiobhyB) [reply docs/how-to-guides/native-mobile/osx-first-time-setup-guide.md]: Noting for posterity that, following https://github.com/WordPress/gutenberg/pull/36007 and the comment here https://github.com/WordPress/gutenberg/pull/36007#pullrequestreview-791037693, it doesn't seem like separate instructions for M1 are now necessary. 🎉

**Review by SiobhyB — APPROVED** [review]
This all looks good to me @gwwar, pending the updated references to the images you've suggested in your previous comment :)

**Comment by gwwar** [comment]
Thanks @SiobhyB! I've also added image assets in https://github.com/WordPress/gutenberg/pull/35529/commits/04b03b2c70537331c631f32e655e495df58c99e7

cc @mkaz did you have any recommendations on formatting/readme location or does this look ok to land?

**Comment by mkaz** [comment]
Looks great! Thanks so much for the updates, it'll make it easier to get more people started and involved 👍 

Can we rename the file to: `osx-setup-guide.md` ? 
Since it still applies even if it's not your first time, makes for a slightly cleaner URL.

Also, you will need to add an entry to `docs/toc.json` (see existing for syntax) and run `npm run docs:build` and commit the generated manifest.json file.

I suppose it doesn't quite matter where in the How to Guides section of the toc.json list. I'm trying to get it in alphabetical order (https://github.com/WordPress/gutenberg/pull/35904). I'll rebase and fix that ticket if yours goes in first.

**Comment by gwwar** [comment]
Ah I knew I must have been missing something! Thanks @mkaz! updated in fbb0285

**Comment by mkaz** [comment]
I just realized, since it's a new directory you may also want a README.md, I'm not really sure what it will do without. I'm ok with trying it as it is and we can add a README if needed.

**Comment by gwwar** [comment]
Sure we can try as is, and see if we need to add an additional README.md as a follow up.

**Comment by gwwar** [comment]
Going to rebase to see if unit test issues resolve themselves

**Comment by gwwar** [comment]
Thanks for the reviews @SiobhyB @mkaz !
